File: golang/construct_binary_tree.go
func buildTree(preorder []int, inorder []int) *TreeNode {
	return helper(preorder, inorder, 0, 0, len(inorder)-1)
}

func helper(preorder, inorder []int, pre, start, end int) *TreeNode {
	// base case
	if start > end || pre >= len(preorder) {
		return nil
	}

	root := &TreeNode{Val: preorder[pre]}

	// find index inside of inorder
	var pivot int
	for i := start; i <= end; i++ {
		if root.Val == inorder[i] {
			pivot = i
		}
	}

	// next root candidates are determined from current pre index + offset, left always pre+1, right considers offset of left elements from inorder pivot
	// offset right for new head by num elements to left of pivot
	offset := pivot - start
	root.Left = helper(preorder, inorder, pre+1, start, pivot-1)
	root.Right = helper(preorder, inorder, pre+offset+1, pivot+1, end)

	return root
}

File: typescript/construct_binary_tree.ts
function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
	// preorder is parent left right, inorder is left parent right
	// root is guaranteed to be first value of preorder
	// find root as the pivot inside inorder => find left/right head from offset
	return helper(preorder, inorder, 0, 0, inorder.length - 1)
}

function helper(preorder: number[], inorder: number[], pre: number, start: number, end: number): TreeNode {
	// base case
	if (start > end) {
		return null // marks child as null
	}

	// find the pivot and offset
	let root = preorder[pre]
	let pivot = 0
	for (let i = start; i <= end; i++) {
		if (inorder[i] == root) {
			pivot = i
			break
		}
	}
	let offset = pivot - start + 1 // represents number of elements in left subtree => used to find right head

	// define node
	let node = new TreeNode(root, null, null)
	node.left = helper(preorder, inorder, pre + 1, start, pivot - 1)
	node.right = helper(preorder, inorder, pre + offset, pivot + 1, end)

	return node
}
